### DISCUSSION TITLE
Dynamically modifying keybindings/configuration?

### DISCUSSION THREAD

**Discussion by Lockszmith-GH** [zellij-org/zellij#2142] [Q&A] [answered]
Question:
Is there a way to change keybindings from a running session?
Is there a way to have the setup conditionally choose a block of configuration based on environment variable?

Where I'm coming from / what I'm trying to do:
I have phased out tmux and replaced it with Zellij - a lot of superior options, really happy about it.
Since at times I use Zellij in a nested state, I wanted to bind the lock key differently depending on where I'm at.
I identify whether I'm nested or not vi an environment-variable.

I can have multiple sets of conifgs where the only difference is the lock/unlock keybinding, but that means that I'll need to keep configuration changes in sync.

I was hoping to find a solution where keybindings are assigned dynamically, but I can't figure it out on my own.
I saw that layouts can have keybindings, but a layout file without a layout isn't something I can load.

Does anyone have a suggestion, or has attempted something like this?

**Comment by imsnif** [comment]
Hey, so I'll start by saying that right now we unfortunately do not have a good solution for this. I also suffer from this problem when developing Zellij (I'm forced to open my development version in a different terminal window than the one I'm developing in - very sad!).

The way I was thinking of fixing this is to have Zellij aware of whether it's nested or not. And then once you lock while focused on a pane with a nested session, it will first lock your outer session, then your inner session, then unlock the outer, then unlock the inner. Of course also working with multiple layers of nesting.

What do you think, will this future feature solve your problem?

**Reply by Lockszmith-GH** [reply]
That can work and will save new-comers some grief if they find themselves in that situation - as a solution will be presented.

But I would rather setup a keybinding for nested sessions based on the environment - this way I can force a specific behavior via command line.
Also, how would you detect this across an SSH connection?

I really like the layout dynamic loading option, if its possible to use that as a dynamic config loader that would make things very flexible.

**Reply by imsnif** [reply]
Changing keybindings configuration at runtime is a bit further down the line for us, I'm afraid. Definitely planned (I want this to also be part of the UI - you click on one of the hints and are prompted to rebind it), but we've got a lot of things to develop before this. Hopefully this will happen some time this year!

**Reply by Lockszmith-GH** [reply, marked as answer]
so sounds like overlapping configurations, or conditional rules might be a solution here.

For now, I'm going to use a template file with `envsubst` and initiate the configuration based on that template.

Below is the exact details if anyone else is interested in understanding what I'm trying to simplify.

## Auto start code
This is how I start zellij today - NOTE the `$USER_IS_HOME` test:
```
ZELLIJ_CFG_PATH="$HOME/.config/zellij/config"
ZELLIJ_CONFIG_FILE="${ZELLIJ_CFG_PATH}.ctrl-g.kdl"
[[ -z "$USER_IS_HOME" ]] || ZELLIJ_CONFIG_FILE="${ZELLIJ_CFG_PATH}.ctrl-b.kdl"
. <({
    zellij setup --generate-auto-start bash \
    | sed -E 's/attach -c/attach --create "${SUDO_USER:-$USER}@$(hostname)"/; '
})
```

## `edit-zellij` function
This is how I edit zellij config - I edit the template and have it generate the actual config files:
```
function edit-zellij() {
    ZELLIJ_CFG_PATH="$HOME/.config/zellij/config"
    if [[ ! -r "${ZELLIJ_CFG_PATH}.kdl.template" ]]; then
        mkdir -p "$(dirname "${ZELLIJ_CFG_PATH}")"
        zellij setup --dump-config \
            | sed -E '
                s|keybinds\W*\{|keybinds clear-defaults=true {|;
                s|^(\W*)(\w*.*tmux.*)$|\1/-\2|;
                s|(bind "Ctrl q" \{ Quit; \})|/-\1|;
                s/"Ctrl g" (\{ SwitchToMode "(Normal|Locked)")/"${ZELLIJ_LOCK_KEYBIND}" \1/;
                s/"Ctrl t" (\{ SwitchToMode "(Normal|Tab)")/"Ctrl u" \1/;
            ' \
            > "${ZELLIJ_CFG_PATH}.kdl.template"
    fi

    [[ "$1" == "--setup" ]] || vi "${ZELLIJ_CFG_PATH}.kdl.template"

    ZELLIJ_LOCK_KEYBIND="Ctrl b" envsubst < "${ZELLIJ_CFG_PATH}.kdl.template" > "${ZELLIJ_CFG_PATH}.ctrl-b.kdl"
    ZELLIJ_LOCK_KEYBIND="Ctrl g" envsubst < "${ZELLIJ_CFG_PATH}.kdl.template" > "${ZELLIJ_CFG_PATH}.ctrl-g.kdl"

    [[ -r "${ZELLIJ_CFG_PATH}.kdl" ]] \
    || ln -sr "${ZELLIJ_CFG_PATH}.ctrl-g.kdl" "${ZELLIJ_CFG_PATH}.kdl"
}
```
If the template does not exist, it is initialized from the default setup applying some defaults of my own.
Most importantly, it replaces the `Ctrl g "{ SwitchToMode` code to one with the template variable.

After editing or `--setup` arg, it generates the different `.kdl` files, it also links the default for good measures.

## Initial Setup
This is how I install zellij on linux/bash - it calls the edit code above, with `--setup` as an argument:
```
. <(curl -sL zellij.dev/launch | sed 's/curl /curl -s /g; s/exit/return/g') -V \
|| eval $( error 1 "Failed to download zellij" )

mkdir -p $HOME/.local/bin
mv $dir/zellij $HOME/.local/bin && rm -fR /tmp/zellij*

edit-zellij --setup
```
